src/mpid/ch4/netmod/ofi/ofi_gpu_pipeline.c

```diff
@@ -278,45 +278,77 @@ int MPIDI_OFI_gpu_pipeline_recv(MPIR_Request * rreq,
     MPIDI_OFI_REQUEST(rreq, noncontig.pack.count) = count;
     MPIDI_OFI_REQUEST(rreq, noncontig.pack.datatype) = datatype;
 
-    struct recv_alloc *p;
-    p = MPL_malloc(sizeof(*p), MPL_MEM_OTHER);
-    MPIR_Assert(p);
+    mpi_errno = MPIR_Async_things_add(recv_init_alloc_poll, rreq);
 
-    p->rreq = rreq;
-    p->n_chunks = -1;   /* it's MPIDI_OFI_EVENT_RECV_GPU_PIPELINE_INIT */
+    return mpi_errno;
+}
 
-    mpi_errno = MPIR_Async_things_add(recv_alloc_poll, p);
+static int recv_init_alloc_poll(MPIR_Async_thing * thing)
+{
+    MPIR_Request *rreq = MPIR_Async_thing_get_state(thing);
 
-    return mpi_errno;
+    int ret = issue_recv_alloc(rreq, true /* is_init */);
+    if (ret) {
+        return MPIR_ASYNC_THING_DONE;
+    }
+
+    return MPIR_ASYNC_THING_NOPROGRESS;
 }
 
+/* ---- */
+struct recv_chunk_alloc {
+    MPIR_Request *rreq;
+    int n_chunks;
+    int issued_chunks;
+};
+
+static int recv_chunk_alloc_poll(MPIR_Async_thing * thing);
+
 /* this is called from recv_event */
-static int start_recv_chunk(MPIR_Request * rreq, int idx, int n_chunks)
+static int start_recv_chunk(MPIR_Request * rreq, int n_chunks)
 {
     int mpi_errno = MPI_SUCCESS;
 
-    struct recv_alloc *p;
+    struct recv_chunk_alloc *p;
     p = MPL_malloc(sizeof(*p), MPL_MEM_OTHER);
     MPIR_Assert(p);
 
     p->rreq = rreq;
     p->n_chunks = n_chunks;
+    p->issued_chunks = 0;
 
-    mpi_errno = MPIR_Async_things_add(recv_alloc_poll, p);
+    mpi_errno = MPIR_Async_things_add(recv_chunk_alloc_poll, p);
 
     return mpi_errno;
 }
 
-static int recv_alloc_poll(MPIR_Async_thing * thing)
+static int recv_chunk_alloc_poll(MPIR_Async_thing * thing)
 {
-    struct recv_alloc *p = MPIR_Async_thing_get_state(thing);
+    struct recv_chunk_alloc *p = MPIR_Async_thing_get_state(thing);
     MPIR_Request *rreq = p->rreq;
 
     /* arbitrary threshold */
     if (MPIDI_OFI_REQUEST(rreq, pipeline_info.recv.num_inrecv) > 1) {
```

Comment:
just so i'm understanding, the number of outstanding pipeline recvs is capped at 1?

Reply:
Yea. My thinking was that we probably don't want thousands of chunk recvs waiting since they all match the same signature. Thus we need an arbitrary threshold. `1` is the minimum, but we probably can use something bigger, like `10`.

Reply:
CXI is sensitive to expected vs. unexpected recvs. Did the previous code post everything at once? I'm concerned this will negatively affect performance if the number is too small.

Reply:
I think previously will keep posting until running out of genq buffer or fi_recv return error, then it goes into one at a time. I think it is equivalent to no limit here. What do you suggest? Remove the limit altogether, or set it to some reasonable number e.g. `100`?

Reply:
I'm leaning towards no limit.